plugins/inputs/elasticsearch_query/aggregation_query.go

```diff
@@ -0,0 +1,184 @@
+	var ftype string
+	mapMetricFields := make(map[string]string)
+
+	if e.ESClient == nil {
+		err := e.connectToES()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// TODO: check if it is possible to improve this
+	for _, metricField := range aggregation.MetricFields {
+
+		resp, err := e.ESClient.GetFieldMapping().Index(aggregation.Index).Field(metricField).Do(ctx)
+
+		if err != nil {
+			return mapMetricFields, err
+		}
+
+		for _, index := range resp {
+			for _, _type := range index.(map[string]interface{})["mappings"].(map[string]interface{}) {
+				for _, field := range _type.(map[string]interface{}) {
+					fname := field.(map[string]interface{})["full_name"].(string)
+					for _, fieldType := range field.(map[string]interface{})["mapping"].(map[string]interface{}) {
+						ftype = fieldType.(map[string]interface{})["type"].(string)
+					}
+					mapMetricFields[fname] = ftype
+				}
+			}
+		}
+	}
+
+	return mapMetricFields, nil
+}
+
+func (e *ElasticsearchQuery) buildAggregationQuery(mapMetricFields map[string]string, aggregation Aggregation) ([]aggregationQueryData, error) {
+
+	aggregationQueryList := make([]aggregationQueryData, (len(aggregation.Tags) + len(aggregation.MetricFields)))
+
+	// create one aggregation per metric field found & function defined for numeric fields
+	for k, v := range mapMetricFields {
+		switch v {
+		case "long":
+		case "float":
+		case "integer":
+		case "short":
+		case "double":
+		case "scaled_float":
+		default:
+			continue
+		}
+
+		agg, err := e.getFunctionAggregation(aggregation.MetricFunction, k)
+		if err != nil {
+			return nil, err
+		}
+
+		aggregationQuery := aggregationQueryData{
+			aggKey: aggKey{
+				measurement: aggregation.MeasurementName,
+				function:    aggregation.MetricFunction,
+				field:       k,
+				name:        strings.Replace(k, ".", "_", -1) + "_" + aggregation.MetricFunction},
+			isParent:    true,
+			aggregation: agg,
+		}
+
+		aggregationQueryList = append(aggregationQueryList, aggregationQuery)
+
+	}
+
+	for _, term := range aggregation.Tags {
+		aggregationQueryList = e.getTermsAggregation(aggregation.MeasurementName, term, aggregation.IncludeMissingTag, aggregation.MissingTagValue, aggregationQueryList)
+	}
+
+	return aggregationQueryList, nil
+}
+
+func (e *ElasticsearchQuery) getFunctionAggregation(function string, aggfield string) (elastic.Aggregation, error) {
+
+	var agg elastic.Aggregation
+
+	switch function {
+	case "avg":
+		agg = elastic.NewAvgAggregation().Field(aggfield)
+	case "sum":
+		agg = elastic.NewSumAggregation().Field(aggfield)
+	case "min":
+		agg = elastic.NewMinAggregation().Field(aggfield)
+	case "max":
+		agg = elastic.NewMaxAggregation().Field(aggfield)
+	// TODO:
+	// case "percentile":
+	// 	agg = elastic.NewPercentilesAggregation().Field(aggfield)
+	default:
+		return agg, fmt.Errorf("Aggregation function %s not supported", function)
```

Comment:
In case you error please return `nil` for the value if possible. This way you directly see that you cannot rely on the value and do not need to trace back the variable.